CUDADataFormats/HGCal/src/RecHitGPUProduct.cc

```diff
@@ -0,0 +1,12 @@
+#include "CUDADataFormats/HGCal/interface/RecHitGPUProduct.h"
+
+RecHitGPUProduct::RecHitGPUProduct(const uint32_t& nhits, const cudaStream_t& stream) : nhits_(nhits) {
+  stride_ = ((nhits - 1) / 32 + 1) * 32;  //align to warp boundary
```

Comment:
It is possible I did not understand your comment, but I will try to justify my choice. A warp is, AFAIK, to this date, always a group of 32 threads. In order to guarantee coalesced memory access, I have to make sure each variable stored within my SoA has enough allocated memory to store a multiple of 32 *in units of number of hits* (as the variable ```stride_``` above, which should, by the way, be called ```padding_``` or related instead). This approach is independent of the type size of the variable in the SoA being considered, as the *padded* (the first multiple of 32 larger or equal to the number of hits) number of hits is then multiplied by the size of each type (see ```CUDADataFormats/HGCal/src/RecHitGPUProduct.cc```).

When a load operation is issued, as in Slides 17 and 18 of [this presentation](https://bluewaters.ncsa.illinois.edu/liferay-content/document-library/Documentation%20Documents/Workshops/Advanced%20User%20Workshop%20Oct%202014/NCSA02_Fundamental_CUDA_Optimization.pdf), the required memory addresses and corresponding cache lines are accessed. In my case, assuming a 4-byte type, each warp will request 128 bytes, which might correspond to a single cache line depending on the specific GPU being used. Even if the cache line just a fraction of 128 byes (32 or 64, for instance), I will be using the bus at a 100% efficiency.

I am sorry if all of this is obvious; I am struggling to understand what you were referring to with the mentioned 128 bytes alignment.

As a side comment, do you think I should replace "32" by some global cuda variable, as discussed [here](https://stackoverflow.com/questions/36047035/when-should-i-use-cudas-built-in-warpsize-as-opposed-to-my-own-proper-constant)?

Reply:
> In order to guarantee coalesced memory access, I have to make sure each variable stored within my SoA has enough allocated memory to store a multiple of 32 _in units of number of hits_ (as the variable `stride_` above, which should, by the way, be called `padding_` or related instead). This approach is independent of the type size of the variable in the SoA being considered, as the _padded_ (the first multiple of 32 larger or equal to the number of hits) number of hits is then multiplied by the size of each type (see `CUDADataFormats/HGCal/src/RecHitGPUProduct.cc`).

Thanks for the clarification, I agree by using the `stride_` as a unit of 4-byte elements leads to the correct memory access pattern. But it is difficult to come to that conclusion here, because the pointer assignments (i.e. the begin addresses of each column) are done in a different file in a separate package (`RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalProducerMemoryWrapper.cc` IIUC).

Reply:
> As a side comment, do you think I should replace "32" by some global cuda variable, as discussed [here](https://stackoverflow.com/questions/36047035/when-should-i-use-cudas-built-in-warpsize-as-opposed-to-my-own-proper-constant)?

Possibly. Using `warpSize` would reduce the effort to use e.g. AMD GPUs (of which some have "warps" of 64 threads), but to get there, we'd hopefully use some higher-level portability tool (rather than HIP directly), which would imply reviewing and adjusting all CUDA code anyway. There are also many cases in the pixel code (at least) that use hardcoded `32`.